### PR TITLE
Add aten::cdist_forward

### PR DESCRIPTION
**Pull request by huaiyuzh**
support cdist_forward

### CODE REVIEW

File: src/aten/sycl/DistanceKernel.cpp [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+};
+
+// General p norm
+template <typename scalar_t>
+struct dists_p {
+  static void inc(scalar_t& agg, const scalar_t diff, const scalar_t p) {
+    // TODO:
+    // Here had an unknown bug which affected other code segments
+    // unexpectedly. See Jira:
+    // https://jira.devtools.intel.com/browse/PYTORCHDGQ-958 for details Below
```
Comment (dvrogozh) [inline src/aten/sycl/DistanceKernel.cpp:94]: Which bug are you talking about? The link you provide is intel internal link and is not helpful to non-intel people who might review/try out this code. Can you, please, create github issue instead with the description?

File: src/aten/sycl/DistanceKernel.cpp [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+  agg = (local_id < sg_num) ? local_shared_mem[lane_id] : (scalar_t)0.0f;
+  if (0 == sg_id) {
+    agg = subgroup_reduce_agg<scalar_t, F, nd_item>(item, agg, sg_size);
+  }
+
+  return agg;
+}
+
+template <typename scalar_t, typename F, int p_type, typename accscalar_t>
+struct CdistForwardKernelImplFunctor  : public __SYCL_KER_CONFIG_CONVENTION__ {
```
Comment (fengyuan14) [inline src/aten/sycl/DistanceKernel.cpp:194]: One more space.

File: src/aten/sycl/DistanceKernel.cpp [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+    result = at::zeros_symint(output_shape, x1.options());
+  } else if (p == 2 && (mode == 1 || (mode == 0 && (r1 > 25 || r2 > 25)))) {
+    Tensor dist = (expand_batch_product == 1)
+        ? at::_euclidean_dist(x1, x2)
+        : at::_euclidean_dist(x1_expanded, x2_expanded);
+    result = dist.view_symint(output_shape);
+  } else {
+    result = at::empty_symint(output_shape, x1.options());
+    cdist_kernel_impl(result, x1_expanded, x2_expanded, p);
+  }
```
Comment (fengyuan14) [inline src/aten/sycl/DistanceKernel.cpp:442]: Remove to op level

File: src/aten/sycl/DistanceKernel.cpp [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+}
+
+void cdist_kernel_impl(Tensor& result, const Tensor& x1_expanded, const Tensor& x2_expanded, double p){
+  const int64_t r1 = x1_expanded.size(-2);
+  const int64_t r2 = x2_expanded.size(-2);
+  const int64_t m = x1_expanded.size(-1);
+
+  AT_DISPATCH_FLOATING_TYPES(
+        x1_expanded.scalar_type(),
+        "cdist_forward_sycl",
```
Comment (fengyuan14) [inline src/aten/sycl/DistanceKernel.cpp:314]:
```suggestion
        "cdist_forward_xpu",
```

File: src/aten/sycl/DistanceKernel.cpp [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+  accscalar_t p_val_;
+  scalar_t* out_data_;
+  scalar_t* x1_data_;
+  scalar_t* x2_data_;
+  sycl_local_acc_t<scalar_t, 1> shared_;
+  const int64_t wgroup_size_;
+};
+
+template <typename scalar_t, typename F, int p_type>
+static void cdist_forward_kernel_impl(
```
Comment (fengyuan14) [inline src/aten/sycl/DistanceKernel.cpp:272]: `launch_cdist_forward_kernel`

File: src/aten/sycl/DistanceKernel.cpp [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+    if (local_id == 0) {
+      out_ptr[group_id] = F::finish(agg, p_val_);
+    }
+  }
+
+  void sycl_ker_config_convention(sycl::handler& cgh) {
+    shared_ = sycl_local_acc_t<scalar_t>(wgroup_size_, cgh);
+  }
+
+  CdistForwardKernelImplFunctor(
```
Comment (fengyuan14) [inline src/aten/sycl/DistanceKernel.cpp:231]: `CdistForwardKernelFunctor`

File: src/aten/sycl/DistanceKernel.cpp [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+      r_size,
+      l1_size,
+      l2_size,
+      p_val,
+      out_data,
+      x1_data,
+      x2_data,
+      wgroup_size);
+  auto& sycl_queue = getCurrentSYCLQueue();
+  sycl_kernel_submit(ngroups * wgroup_size, wgroup_size, sycl_queue, kfn);
```
Comment (fengyuan14) [inline src/aten/sycl/DistanceKernel.cpp:304]:
```suggestion
  sycl_kernel_submit(ngroups * wgroup_size, wgroup_size, queue, kfn);
```

File: test/xpu/run_test_with_only.py [hunk 8]
```diff
@@ -58,6 +58,31 @@ def launch_test(test_case, skip_list=None, exe_list=None):
 # test_comparison_utils
 res += launch_test("test_comparison_utils_xpu.py")
 
+#test_meta
+execute_list = (
+    "test_dispatch_symbolic_meta_outplace_all_strides_cdist_xpu_float32",
```
Comment (fengyuan14) [inline test/xpu/run_test_with_only.py:63]: Can only run float32? Can we run half?

File: test/xpu/run_test_with_skip.py [hunk 9]
```diff
@@ -941,6 +941,12 @@ def launch_test(test_case, skip_list=None, exe_list=None):
     "test_conj_view_to_sparse_xpu_complex64",
     "test_neg_conj_view_to_sparse_xpu_complex128",
     "test_neg_view_to_sparse_xpu_float64",
+
+    # AssertionError: False is not true : Argument 0 during forward call unexpectedly materializes. 
+    # Either set `supports_cow_input_no_materialize_forward=False` in this operation's OpInfo, 
+    # add the arg to the OpInfo's `allow_cow_input_materialize_forward` list, 
+    # or change the implementation to avoid materialization.
+    "test_cow_input_cdist_xpu_float32",
```
Comment (fengyuan14) [inline test/xpu/run_test_with_skip.py:949]: It should be a case of `test_ops::TestCompositeCompliance`. But the suite is not enabled here, but enabled in fine gran cases? Why add skip here?

Comment (fengyuan14) [reply test/xpu/run_test_with_skip.py:949]: I see, you add the suite here.

**Comment by fengyuan14** [comment]
Please add fine gran cases.

File: test/xpu/run_test_with_only.py (hunk 8 above)
Comment (huaiyuzh) [reply test/xpu/run_test_with_only.py:63]: only float32 case

File: src/ATen/native/xpu/sycl/DistanceKernels.cpp [hunk 10]
```diff
@@ -0,0 +1,380 @@
+#pragma clang diagnostic push
+#pragma GCC diagnostic push
+// Avoid SYCL compiler return-type error
+#pragma clang diagnostic ignored "-Wreturn-type"
+#pragma GCC diagnostic ignored "-Wreturn-type"
+
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/DistanceKernels.cpp:6]: Why these?

File: src/ATen/native/xpu/sycl/DistanceKernels.h [hunk 11]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+#include <ATen/ATen.h>
+
+namespace at::native::xpu {
+void cdist_kernel_impl(
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/DistanceKernels.h:5]: cdist_kernel

**Comment by fengyuan14** [comment]
Silu random failure. Retrigger failed preci flow.

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**